### ISSUE TITLE
java.lang.ClassCastException when deleting a branch

### ISSUE THREAD

**Issue by latitude37s** [sford/yet-another-commit-checker#202]
Immediately after upgrading to v1.21 we starting seeing java.lang.ClassCastException errors in atlassian-bitbucket.log. These seem to occur when trying to delete a branch.

We have author and email checks configured in the YACC global settings.

We have since downgraded back to v1.19.

```
Error calling com.isroot.stash.plugin.YaccGlobalHook.preUpdate (com.isroot.stash.plugin.yacc:yaccGlobalHook)
java.lang.ClassCastException: com.atlassian.bitbucket.event.branch.BranchDeletionHookRequest cannot be cast to com.atlassian.bitbucket.event.branch.BranchCreationHookRequest
        at com.isroot.stash.plugin.YaccGlobalHook.handleGlobalHook(YaccGlobalHook.java:93)
        at com.isroot.stash.plugin.YaccGlobalHook.preUpdate(YaccGlobalHook.java:74)
        at com.atlassian.stash.internal.hook.repository.DefaultRepositoryHookService.preUpdate(DefaultRepositoryHookService.java:860)
        at com.atlassian.stash.internal.hook.repository.DefaultRepositoryHookService.lambda$preUpdate$4(DefaultRepositoryHookService.java:458)
        at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
        at com.atlassian.stash.internal.hook.repository.DefaultRepositoryHookService.preUpdate(DefaultRepositoryHookService.java:437)
        at com.atlassian.plugin.util.ContextClassLoaderSettingInvocationHandler.invoke(ContextClassLoaderSettingInvocationHandler.java:26)
        at org.eclipse.gemini.blueprint.service.importer.support.internal.aop.ServiceInvoker.doInvoke(ServiceInvoker.java:56)
        at org.eclipse.gemini.blueprint.service.importer.support.internal.aop.ServiceInvoker.invoke(ServiceInvoker.java:60)
        at org.eclipse.gemini.blueprint.service.util.internal.aop.ServiceTCCLInterceptor.invokeUnprivileged(ServiceTCCLInterceptor.java:70)
        at org.eclipse.gemini.blueprint.service.util.internal.aop.ServiceTCCLInterceptor.invoke(ServiceTCCLInterceptor.java:53)
        at org.eclipse.gemini.blueprint.service.importer.support.LocalBundleContextAdvice.invoke(LocalBundleContextAdvice.java:57)
        at com.atlassian.bitbucket.internal.branch.DefaultBranchService.deleteBranch(DefaultBranchService.java:97)
        at com.atlassian.bitbucket.internal.branch.rest.BranchUtilsResource.deleteBranch(BranchUtilsResource.java:146)
        at com.atlassian.applinks.core.rest.context.ContextFilter.doFilter(ContextFilter.java:24)
        at com.atlassian.applinks.core.rest.context.ContextFilter.doFilter(ContextFilter.java:24)
        at com.atlassian.applinks.core.rest.context.ContextFilter.doFilter(ContextFilter.java:24)
        at com.atlassian.applinks.core.rest.context.ContextFilter.doFilter(ContextFilter.java:24)
        at com.atlassian.applinks.core.rest.context.ContextFilter.doFilter(ContextFilter.java:24)
        at com.atlassian.analytics.client.filter.UniversalAnalyticsFilter.doFilter(UniversalAnalyticsFilter.java:92)
        at com.atlassian.analytics.client.filter.AbstractHttpFilter.doFilter(AbstractHttpFilter.java:39)
        at com.atlassian.bitbucket.internal.xcode.web.XcodeUserAgentFilter.doFilter(XcodeUserAgentFilter.java:36)
        at com.atlassian.stash.internal.spring.lifecycle.LifecycleJohnsonServletFilterModuleContainerFilter.doFilter(LifecycleJohnsonServletFilterModuleContainerFilter.java:42)
        at com.atlassian.plugin.connect.plugin.auth.scope.ApiScopingFilter.doFilter(ApiScopingFilter.java:81)
        at com.atlassian.stash.internal.spring.lifecycle.LifecycleJohnsonServletFilterModuleContainerFilter.doFilter(LifecycleJohnsonServletFilterModuleContainerFilter.java:42)
        at com.atlassian.stash.internal.spring.security.StashAuthenticationFilter.doFilter(StashAuthenticationFilter.java:85)
        at com.atlassian.stash.internal.web.auth.BeforeLoginPluginAuthenticationFilter.doInsideSpringSecurityChain(BeforeLoginPluginAuthenticationFilter.java:112)
        at com.atlassian.stash.internal.web.auth.BeforeLoginPluginAuthenticationFilter.doFilter(BeforeLoginPluginAuthenticationFilter.java:75)
        at com.atlassian.security.auth.trustedapps.filter.TrustedApplicationsFilter.doFilter(TrustedApplicationsFilter.java:94)
        at com.atlassian.oauth.serviceprovider.internal.servlet.OAuthFilter.doFilter(OAuthFilter.java:67)
        at com.atlassian.stash.internal.spring.lifecycle.LifecycleJohnsonServletFilterModuleContainerFilter.doFilter(LifecycleJohnsonServletFilterModuleContainerFilter.java:42)
        at com.atlassian.plugin.connect.plugin.auth.oauth2.DefaultSalAuthenticationFilter.doFilter(DefaultSalAuthenticationFilter.java:69)
        at com.atlassian.plugin.connect.plugin.auth.user.ThreeLeggedAuthFilter.doFilter(ThreeLeggedAuthFilter.java:109)
        at com.atlassian.jwt.internal.servlet.JwtAuthFilter.doFilter(JwtAuthFilter.java:32)
        at com.atlassian.analytics.client.filter.DefaultAnalyticsFilter.doFilter(DefaultAnalyticsFilter.java:38)
        at com.atlassian.analytics.client.filter.AbstractHttpFilter.doFilter(AbstractHttpFilter.java:39)
        at com.atlassian.bitbucket.internal.scm.git.lfs.servlet.filter.GitLfsLockingFilter.doFilter(GitLfsLockingFilter.java:50)
        at com.atlassian.stash.internal.spring.lifecycle.LifecycleJohnsonServletFilterModuleContainerFilter.doFilter(LifecycleJohnsonServletFilterModuleContainerFilter.java:42)
        at com.atlassian.stash.internal.web.auth.BeforeLoginPluginAuthenticationFilter.doBeforeBeforeLoginFilters(BeforeLoginPluginAuthenticationFilter.java:90)
        at com.atlassian.stash.internal.web.auth.BeforeLoginPluginAuthenticationFilter.doFilter(BeforeLoginPluginAuthenticationFilter.java:73)
        at com.atlassian.stash.internal.request.DefaultRequestManager.doAsRequest(DefaultRequestManager.java:89)
        at com.atlassian.stash.internal.hazelcast.ConfigurableWebFilter.doFilter(ConfigurableWebFilter.java:38)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.lang.Thread.run(Thread.java:748)
        ... 290 frames trimmed
```

**Comment by manzonico84** [comment]
Hi Yacc team,

The same thing happens to me. I did the update to the same version a few hours ago.

Can you help us guys, 
Fabio

![image](https://user-images.githubusercontent.com/45595229/49441241-0c25b200-f7ad-11e8-84c6-5e24d8e370b6.png)

**Comment by sford** [comment]
Looks like BBS is sending wrong request for trigger:
```
[INFO] 2018-12-04 07:39:54,297 DEBUG [http-nio-7990-exec-7] admin @1LU7S85x459x345x0 127.0.0.1 "DELETE /rest/branch-utils/1.0/projects/PROJECT_1/repos/07cc5eda-bf0b-4683-a71c-945d8c0258d8/branches HTTP/1.1" com.isroot.stash.plugin.YaccHook preUpdate, class=BranchDeletionHookRequest, trigger=BRANCH_CREATE
```

Trigger is `BRANCH_CREATE`, but `BranchDeletionHookRequest` is being sent. This request should have trigger `BRANCH_DELETE`.

**Comment by sford** [comment]
This is a bug in BBS, looks like it was just fixed: https://jira.atlassian.com/browse/BSERV-11458?jql=text%20~%20%22BranchDeletionHookRequest%22

YACC needs a workaround for this.

**Comment by sford** [comment]
Fixed in 1.22
